### PR TITLE
[abseil] Add compiler C++17.

### PR DESCRIPTION
**Pull request by LilyWangL**
Related issue #8181.

### CODE REVIEW

File: ports/abseil/portfile.cmake [hunk 1]
```diff
@@ -20,6 +20,8 @@ vcpkg_from_github(
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}
     PREFER_NINJA
+    OPTIONS
+        -DCMAKE_CXX_STANDARD=17
```
Comment (Neumann-A) [inline ports/abseil/portfile.cmake]: Now people with c++11 will have problems. You can add c++17 as a feature to abseil or simple use ${ABSEIL_OPTIONS} and people can set ABSEIL_OPTIONS to any value they like

**Comment by LilyWangL** [comment]
@jweg Could you test this?
Thanks.

File: ports/abseil/portfile.cmake [hunk 2]
```diff
@@ -17,9 +17,16 @@ vcpkg_from_github(
         fix-config.patch
 )
 
+set(CMAKE_CXX_STANDARD  )
+if("c++17" IN_LIST FEATURES)
```
Comment (jweg) [inline ports/abseil/portfile.cmake]: This does fix my crash, thank you. Will that still work if user code wants to build as C++20? Can't test this right now, my GCC here is not new enough. I still wouldn't mind a way to set this in the triplet so the configuration is in one place, but this works for me.

**Review by cbezault — COMMENTED** [review]
Just from looking at this from a high level I think this should be a triplet setting.
By default your whole project should be built under the same c++ standard.

**Comment by cbezault** [comment]
I've thought of this harder and I think we definitely need to come up with a way to handle this in the triplet.

File: ports/abseil/portfile.cmake (hunk 2 above)
Comment (cbezault) [reply ports/abseil/portfile.cmake]: This should work with C++20, unless they have 17 -> 20 abi breaking changes.

**Comment by jweg** [comment]
So what are the alternatives here?
1. Adding to the compiler options (current way) is imo the worst one -- even if we could enforce that they be last on the command line (that's pretty much why it fails with abseil) we'd rely on later options overwriting previous ones. Also swooping in after the port's processing and change the compiler options is asking for trouble: the port might want to make other decisions based on the language standard to keep everything consistent. It should not be skipped like that. 
2. Inject variables like CMAKE_CXX_STANDARD into the variable namespace (or cache?) -- currently this doesn't seem to work from the triplet. Should it? I feel that this would be too little. Not all ports wrap a CMake build system and then they'd end up awkwardly parsing CMake-specific variables to interpret them in the way the wrapped build system requires.
3. The solution in this PR (special handling of C++17 as a port feature) might feel like a bandaid at first, but I'm not so sure... The port knows whether a feature (in the general sense) needs special treatment and it can declare them with strong, specific semantics with port features. The only thing that makes this awkward is that triplets and ports are independent. This is only to be expected since we get overlap between triplet territory and port features once you start using features for more than strictly selecting subsets of the wrapped library/tool. I think this is a perfectly fine development. I feel we could embrace that with a little change: make it possible to list features per port in the triplet. I see value in having the configuration in one place for the user.

Those thoughts aside, while we don't have anything else, could you please merge this pull request? abseil with C++17+ continues to be a sigsegv mine field with vcpkg without it.

**Comment by PhoebeHui** [comment]
@LilyWangL, could you resolve the conflicts?

I concerned if it's a good choice for us for adding C++17 as a feature. @vicroms, what do you think?

**Comment by dan-shaw** [comment]
Sorry for this taking a while. I'm okay with merging this PR, since we don't have a way to set it in the triplet. 

@LilyWangL can you resolve the conflicts so we can merge this?

**Comment by dan-shaw** [comment]
On second thought and after some discussion with the team, I think its okay if we have it on by default, not a feature.

**Comment by LilyWangL** [comment]
> Sorry for this taking a while. I'm okay with merging this PR, since we don't have a way to set it in the triplet.
> 
> @LilyWangL can you resolve the conflicts so we can merge this?

Done.

**Review by dan-shaw — APPROVED** [review]

**Comment by linquize** [comment]
The feature name cannot have symbols.
```
Error: <unknown>:1:9: invalid character in feature name (must be lowercase, digits, '-', or '*')
   on expression: "abseil[c++17]"
                           ^
```
please change the feature name to something like `cpp17` instead.

**Comment by LilyWangL** [comment]
> The feature name cannot have symbols.
> 
> ```
> vcpkg install abseil[c++17]
> Error: <unknown>:1:9: invalid character in feature name (must be lowercase, digits, '-', or '*')
>    on expression: "abseil[c++17]"
>                            ^
> ```
> 
> please change the feature name to something like `cpp17` instead.

Thanks for your help. It will be fixed in PR #10620

**Pull request merged**